src/Sparrow/Json/ArenaMemoryAllocator.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using NLog;
+
+namespace Sparrow.Json
+{
+    public unsafe class ArenaMemoryAllocator : IDisposable
+    {
+        private byte* _ptrStart;
+        private byte* _ptrCurrent;
+
+        private int _allocated;
+        private int _used;
+
+        private List<IntPtr> _unusedBuffers;
+
+        private bool _isDisposed;
+        private static readonly Logger _log = LogManager.GetLogger(nameof(ArenaMemoryAllocator));
+        
+        public ArenaMemoryAllocator(int initialSize = 1024 * 1024)
+        {
+            _ptrStart = _ptrCurrent = (byte*)Marshal.AllocHGlobal(initialSize).ToPointer();
+            _allocated = initialSize;
+            _used = 0;
+
+            if (_log.IsDebugEnabled)
+                _log.Debug($"ArenaMemoryAllocator was created with initial capacity of {initialSize:#,#;;0} bytes");
+        }
+
+        public AllocatedMemoryData Allocate(int size)
+        {
+            if (_isDisposed)
+                ThrowAlreadyDisposedException();
+
+            if (_used + size > _allocated)
+                GrowArena(size);
+
+            var allocation = new AllocatedMemoryData()
+            {
+                SizeInBytes = size,
+                Address = new IntPtr(_ptrCurrent)
+            };
+
+            _ptrCurrent += size;
+            _used += size;
+
+            if (_log.IsDebugEnabled)
+                _log.Debug($"ArenaMemoryAllocator allocated {size:#,#;;0} bytes");
+
+            return allocation;
+        }
+
+        private void ThrowAlreadyDisposedException()
+        {
+            throw new ObjectDisposedException($"This ArenaMemoryAllocator is already disposed");
+        }
+
+        public void GrowArena(int requestedSize)
+        {
+            while (_used + requestedSize > _allocated)
+            {
+                var newSize = _allocated * 2;
+                if(newSize < _allocated)
```

Comment:
Noticed, fixing it

On Aug 16, 2016 1:31 PM, "Arkadiusz Paliński" <notifications@github.com>
wrote:

> In src/Sparrow/Json/ArenaMemoryAllocator.cs
> <https://github.com/ayende/ravendb/pull/4194#discussion_r74910891>:
>
> > +                _log.Debug($"ArenaMemoryAllocator allocated {size:#,#;;0} bytes");
> > +
> > +            return allocation;
> > +        }
> > +
> > +        private void ThrowAlreadyDisposedException()
> > +        {
> > +            throw new ObjectDisposedException($"This ArenaMemoryAllocator is already disposed");
> > +        }
> > +
> > +        public void GrowArena(int requestedSize)
> > +        {
> > +            while (_used + requestedSize > _allocated)
> > +            {
> > +                var newSize = _allocated * 2;
> > +                if(newSize < _allocated)
>
> If _allocated is 1GB (1073741824 bytes) then in result of * 2 we get
> negative value -2147483648 and throw
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ayende/ravendb/pull/4194/files/78ddcfddf2da8e04f2abfa3820f5b45060980bf5#r74910891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHIs967jhz0VCOaCySiY65YJZg8VfnLks5qgZFqgaJpZM4JhBWU>
> .
>